kbmd/door.c

```diff
@@ -0,0 +1,483 @@
+		(void) bunyan_error(tlog,
+		    "Unable to obtain caller credentials",
+		    BUNYAN_T_INT32, "errno", (int32_t)errno,
+		    BUNYAN_T_STRING, "errmsg", strerror(errno),
+		    BUNYAN_T_END);
+		return (B_FALSE);
+	}
+	uc = td->td_ucred;
+	td->td_ucred_set = B_TRUE;
+
+	pw = getpwuid(ucred_getruid(uc));
+	if (pw != NULL) {
+		if (bunyan_key_add(tlog,
+		    BUNYAN_T_STRING, "user", pw->pw_name,
+		    BUNYAN_T_END) != 0)
+			return (B_FALSE);
+	}
+
+	/*
+	 * We rely on bunyan overwriting an existing key with a new
+	 * value for this.
+	 */
+	if (bunyan_key_add(tlog,
+	    BUNYAN_T_UINT32, "req_pid", (uint32_t)ucred_getpid(uc),
+	    BUNYAN_T_UINT32, "req_uid", (uint32_t)ucred_getruid(uc),
+	    BUNYAN_T_END) != 0) {
+		return (B_FALSE);
+	}
+
+	if (n_desc > 1) {
+		pid_t pid = ucred_getpid(td->td_ucred);
+		uid_t uid = ucred_getruid(td->td_ucred);
+
+		(void) bunyan_error(tlog,
+		    "Unexpected number of descrptors passed",
+		    BUNYAN_T_UINT32, "n_desc", n_desc,
+		    BUNYAN_T_UINT32, "req_pid", (uint32_t)pid,
+		    BUNYAN_T_UINT32, "req_uid", (uint32_t)uid,
+		    BUNYAN_T_END);
+
+		return (B_FALSE);
+	}
+
+	if (n_desc == 0) {
+		(void) strlcpy(td->td_tty, "<none>", sizeof (td->td_tty));
+	} else {
+		int rc;
+
+		if ((dp->d_attributes & DOOR_DESCRIPTOR) == 0) {
+			(void) bunyan_error(tlog,
+			    "Passed unknown attribute in door_desc_t",
+			    BUNYAN_T_INT32, "attr", dp->d_attributes,
+			    BUNYAN_T_END);
+			return (B_FALSE);
+		}
+
+		td->td_errfd = dp->d_data.d_desc.d_descriptor;
+		rc = ttyname_r(td->td_errfd, td->td_tty, sizeof (td->td_tty));
+		if (rc == 0) {
+			(void) strlcpy(td->td_tty, "<none>",
+			    sizeof (td->td_tty));
+		}
+	}
+
+	if (bunyan_key_add(tlog,
+	    BUNYAN_T_STRING, "req_tty", td->td_tty,
+	    BUNYAN_T_END) != 0) {
+		return (B_FALSE);
+	}
+
+	return (B_TRUE);
+}
+
+static tdata_t *
+tdata_alloc(size_t buflen)
+{
+	tdata_t *td = NULL;
+	size_t tdlen = sizeof (*td) + buflen;
+
+	if ((td = umem_zalloc(tdlen, UMEM_DEFAULT)) == NULL)
+		return (NULL);
+
+	if (bunyan_child(blog, &td->td_log, BUNYAN_T_END) != 0) {
+		umem_free(td, tdlen);
+		return (NULL);
+	}
+	tlog = td->td_log;
+
+	td->td_buflen = buflen;
+	td->td_errfd = -1;
+	return (td);
+}
+
+static void
+tdata_free(void *p)
+{
+	if (p == NULL)
+		return;
+
+	tdata_t *td = p;
+	size_t tdlen = 0;
+
+	bunyan_fini(td->td_log);
+	tlog = NULL;
+
+	tdlen = sizeof (*td) + td->td_buflen;
+	umem_free(td, tdlen);
+}
+
+static tdata_t *
+tdata_get(void)
+{
+	tdata_t *td = NULL;
+
+	VERIFY0(thr_keycreate_once(&tdatakey, tdata_free));
+	VERIFY0(thr_getspecific(tdatakey, (void **)&td));
+
+	if (td != NULL)
+		return (td);
+
+	/*
+	 * TODO: Eventually we should make the kbmd door have a private
+	 * preallocated thread pool, with each thread's tdata being
+	 * a required part of the thread creation.
+	 */
+	td = tdata_alloc(door_ret_max);
+	VERIFY3P(td, !=, NULL);
+
+	VERIFY0(thr_setspecific(tdatakey, td));
+	return (td);
+}
+
+uid_t
+req_uid(void)
```

Comment:
Originally, I was thinking access could be controlled via RBAC authorizations, but no time, I'll pull this out and it can be added back later if we want to go that route (an unresolved question with that is 'do we start an illumos.* hierarchy or continue to use the solaris.* one)

Reply:
confirmed